src/source/Common/EcsCredentialProvider.c

```diff
@@ -0,0 +1,290 @@
+
+    *ppAwsCredentials = pEcsCredentialProvider->pAwsCredentials;
+
+CleanUp:
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS parseEcsResponse(PEcsCredentialProvider pEcsCredentialProvider, PCallInfo pCallInfo)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    UINT32 i, resultLen, accessKeyIdLen = 0, secretKeyLen = 0, sessionTokenLen = 0, expirationTimestampLen = 0;
+    INT32 tokenCount;
+    jsmn_parser parser;
+    jsmntok_t tokens[MAX_JSON_TOKEN_COUNT];
+    PCHAR accessKeyId = NULL, secretKey = NULL, sessionToken = NULL, expirationTimestamp = NULL, pResponseStr = NULL;
+    UINT64 expiration, currentTime;
+    CHAR expirationTimestampStr[MAX_EXPIRATION_LEN + 1];
+
+    CHK(pEcsCredentialProvider != NULL && pCallInfo != NULL, STATUS_NULL_ARG);
+
+    resultLen = pCallInfo->responseDataLen;
+    pResponseStr = pCallInfo->responseData;
+    CHK(resultLen > 0, STATUS_ECS_FAILED);
+
+    jsmn_init(&parser);
+    tokenCount = jsmn_parse(&parser, pResponseStr, resultLen, tokens, SIZEOF(tokens) / SIZEOF(jsmntok_t));
+
+    CHK(tokenCount > 1, STATUS_INVALID_API_CALL_RETURN_JSON);
+    CHK(tokens[0].type == JSMN_OBJECT, STATUS_INVALID_API_CALL_RETURN_JSON);
+
+    for (i = 1; i < (UINT32) tokenCount; i++) {
+        if (compareJsonString(pResponseStr, &tokens[i], JSMN_STRING, (PCHAR) "AccessKeyId")) {
+            accessKeyIdLen = (UINT32)(tokens[i + 1].end - tokens[i + 1].start);
+            CHK(accessKeyIdLen <= MAX_ACCESS_KEY_LEN, STATUS_INVALID_API_CALL_RETURN_JSON);
+            accessKeyId = pResponseStr + tokens[i + 1].start;
+            i++;
+        } else if (compareJsonString(pResponseStr, &tokens[i], JSMN_STRING, (PCHAR) "SecretAccessKey")) {
+            secretKeyLen = (UINT32)(tokens[i + 1].end - tokens[i + 1].start);
+            CHK(secretKeyLen <= MAX_SECRET_KEY_LEN, STATUS_INVALID_API_CALL_RETURN_JSON);
+            secretKey = pResponseStr + tokens[i + 1].start;
+            i++;
+        } else if (compareJsonString(pResponseStr, &tokens[i], JSMN_STRING, (PCHAR) "Token")) {
+            sessionTokenLen = (UINT32)(tokens[i + 1].end - tokens[i + 1].start);
+            CHK(sessionTokenLen <= MAX_SESSION_TOKEN_LEN, STATUS_INVALID_API_CALL_RETURN_JSON);
+            sessionToken = pResponseStr + tokens[i + 1].start;
+            i++;
+        } else if (compareJsonString(pResponseStr, &tokens[i], JSMN_STRING, "Expiration")) {
+            expirationTimestampLen = (UINT32)(tokens[i + 1].end - tokens[i + 1].start);
+            CHK(expirationTimestampLen <= MAX_EXPIRATION_LEN, STATUS_INVALID_API_CALL_RETURN_JSON);
+            expirationTimestamp = pResponseStr + tokens[i + 1].start;
+            MEMCPY(expirationTimestampStr, expirationTimestamp, expirationTimestampLen);
+            expirationTimestampStr[expirationTimestampLen] = '\0';
+            i++;
+        }
+    }
+
+    CHK(accessKeyId != NULL && secretKey != NULL && sessionToken != NULL, STATUS_ECS_FAILED);
```

Comment:
Perhaps, we should use a different status code to indicate a parsing of the result error instead of the overall call failure. Will be useful for debugging as well as code processing logic at a higher level.

Reply:
We seem to have a more common error/status if the actual call fails so perhaps this is OK

Reply:
I separated the error code into three error codes. Actually, I agree with you, but I thought this is one code convention in this sdk. So I did this in order to meet the convention.